### ISSUE TITLE
Improve performance of `search` of the relational persistence

### ISSUE THREAD

**Issue by avernet** [orbeon/orbeon-forms#649]
## Goals

* Improve MySQL performance issue caused by search (summary page) running XPath on non-indexed data.
* Make this not MySQL-specific, and rely on the same system in other relational databases, which makes our code more uniform across databases, and should make performance more predictable.

## Queries done by the summary page

* Step 1: search for specific field values:
  * Instead of the current [Oracle `extractValue()`] and [MySQL `extractValue()`]
  * We will query the new table `orbeon_form_data_cont_text_i (data_id, app, form, control, text)`
* Step 2: free text search:
  * Instead of using [`contains()`] in Oracle, [`like`] in MySQL
  * We will query the new table `orbeon_form_data_freetext_i (data_id, app, form, text)`

### Index tables

```sql
create table orbeon_i_current (
    data_id       int           not null,
    document_id   varchar(255)  not null,
    created       timestamp     not null,
    last_modified timestamp     not null,
    username      varchar(255)  not null,
    app           varchar(255)  not null,
    form          varchar(255)  not null,
    index         (data_id),
    index         (app, form),
    foreign key   (data_id) references orbeon_form_data(id)
)   engine = InnoDB;
```

```sql
create table orbeon_i_control_text (
    data_id       int           not null,
    username      varchar(255)  not null,
    app           varchar(255)  not null,
    form          varchar(255)  not null,
    control       varchar(255)  not null,
    pos           int           not null,
    val           text          not null,
    index         (data_id, control),
    index         (app, form, control),
    fulltext      (val),
    foreign key   (data_id) references orbeon_form_data(id)
)   engine = InnoDB;
```

* we use `pos` instead of `position`, as the latter is an [ODBC reserved word].

### Current data table

It isn't clear if we really need a "current data" table for performance reasons. If we do, the DDL and statement to populate it follow.

```sql
create table orbeon_i_data_current (
	data_id int,
    created timestamp,
    last_modified timestamp,
    username varchar(256),
    app varchar(256),
    form varchar(256),
    document_id varchar(256),
    index data_id_i (data_id),
    foreign key (data_id) references orbeon_form_data(id)
)   engine = InnoDB;
```

```sql
insert into orbeon_i_data_current
    select id, created, last_modified, username, app, form, document_id
      from orbeon_form_data
     where (app, form, document_id, last_modified) in
           (
                 select app, form, document_id, max(last_modified) last_modified
                   from orbeon_form_data
               group by app, form, document_id
           )
       and deleted = 'N';
```

### Control name

The `orbeon_i_control_text` table has a `control` column that identifies a given field. For repeated grids and sections, the value of `control` will be the same for all the iterations. So it could almost be the same as the element name in the XML, as we disallow the same control to be used multiple times in the same form. Except:

* That constraint doesn't apply to section templates, which means there could be a control named `gaga` in a section template and another one in the form that uses the section template.
* A custom data model, once supported per #1233, might also have different elements with the same name corresponding to different controls.

A possibility would be to use an XPath-like expression for the control name, e.g. `address/zip`.

### Nested repeats

How do we keep track of the position of controls in nested repeats. E.g. if we have: department / employee / first name, we need to number the departments and employees, and indicate that the value of the control `first_name` corresponds to, say, the first employee in the first department. How can we store this information?

## Next steps

* start handling the `orbeon_form_data` table (we'll deal with `orbeon_form_definition` next)
* write DDL for index tables for MySQL
* write service that populates the tables based on existing forms and data
* write search service in Scala for MySQL using the index tables
* write CRUD service in Scala for MySQL that updates the index tables
* do the same for `orbeon_form_definition`

  [flat view]: http://wiki.orbeon.com/forms/doc/developer-guide/form-runner/oracle-and-mysql-persistence-layers#TOC-Oracle2
  [subquery]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/mysql/crud.xpl#L104
  [dense rank]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/oracle/crud.xpl#L307
  [Oracle `extractValue()`]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/oracle/search.xpl#L112
  [MySQL `extractValue()`]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/mysql/search.xpl#L140
  [`contains()`]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/oracle/search.xpl#L126
  [`like`]: https://github.com/orbeon/orbeon-forms/blob/4.1-ce/src/resources/apps/fr/persistence/mysql/search.xpl#L152
  [foreign keys]: http://dev.mysql.com/doc/refman/5.6/en/create-table-foreign-keys.html
  [ODBC reserved word]: http://msdn.microsoft.com/en-us/library/aa238507(v=sql.80).aspx

**Comment by avernet** [comment]
As a first step, I suggest we do the following, which goes towards the goal above, but is simpler to implement, and provides significant benefits for our customers in the short term, specifically to those who see performance issues with the Form Builder summary page.

We want SQL queries running to produce the summary page, i.e. for the search service, to run without requiring any `extractValue()`. In a first phase, we will do this only for the Form Builder summary page, however keeping in mind that this technique should be extensible so we can use it for other summary pages.

This calls for adding an `id int not null` to `orbeon_form_data`, and 1 index table:

```sql
create table orbeon_i_control_text (
    data_id       int           not null,
    control       varchar(255)  not null,
    val           text          not null,
    foreign key   (data_id) references orbeon_form_data(id)
)
```

Control are identified per the technique used in the search API, i.e. a path. (This could change in the future, but when doing so we'll keep the way used to identify controls in this table in sync with what we are doing in the search API.)

On `INSERT`, if for `app = 'orbeon` and form = 'builder'`, in Scala run the XPath expressions extracting the app name, form name, title, etc, and produce the `INSERT` statements to add this into `orbeon_i_control_text`.

On `search`, if for `app = 'orbeon` and form = 'builder'`, do the search in `orbeon_i_control_text` instead of relying on `extractValue()`.

**Comment by ebruchez** [comment]
Is it more difficult to do it generally rather than just for Form Runner?

**Comment by avernet** [comment]
Not that hard, I think:

- On PUT, instead of just hardcoding the paths, we would need to figure those paths out, based on the form definition, reusing or putting in common the code that does this in the summary page.
- Handle controls in repeated grids/sections, i.e. cases where a control can have multiple values.
- Maybe some special handling will be required typed values (e.g. dates) and selection controls.

There will definitely be a way to get from A to B.

**Comment by avernet** [comment]
- [ ] Consider adding indexes on `orbeon_i_current` similar to those we have on `orbeon_form_data`.
- [ ] Add `data_id` to `orbeon_form_data_attach` so we never have to do a "search" in that table.
- [ ] Sync DDL in `oracle-4_10-to-2016.2.sql` with DDL in `oracle-2016.2.sql`.
- [ ] Amend persistence unit tests to check upgrading the database leads to the same table as if we're starting from scratch.

**Comment by avernet** [comment]
Next:

- [ ] Support drafts in search

**Comment by avernet** [comment]
We are now code complete for Oracle. We still need to:

- Support other relational databases, including writing the DDL for those databases, test our existing code, and making changes where necessary. We would like to do this while writing tests for the existing functionality, which requires #2807 and #2808.
- Provide a user-facing UI for reindexing. Reindexing could be triggered from the home page in admin view, and should show a progress indicator.
- Documentation, in particular explaining what needs to be done when upgrading.

**Comment by avernet** [comment]
We want to add a "reindex everything" action to the Form Runner home page admin view. Currently, it shows 3 buttons:

![screen shot 2016-06-13 at 9 33 42 am](https://cloud.githubusercontent.com/assets/106961/16015311/f0647834-3149-11e6-8e4a-bc48601947f2.png)

All the operations act on the selected form(s). Adding a "reindex" in "operations" would make sense if we want to do such an operation for specific form(s), but a full reindex mixed with other form-specific operations would be confusing. So for now, we'll tentatively add a "Reindex database(s)" button.

**Comment by avernet** [comment]
On the Form Runner home page, now we initially have a button admins can click to start the reindexing:

![screen shot 2016-06-17 at 10 14 10 am](https://cloud.githubusercontent.com/assets/106961/16158891/81af050c-3474-11e6-9b0a-a1b6f916cc64.png)

Once clicked, the reindexing progress shows:

![screen shot 2016-06-17 at 10 14 28 am](https://cloud.githubusercontent.com/assets/106961/16158908/93a87bb2-3474-11e6-9d41-b1fa8da2adcd.png)

If admins click on the button to stop reindexing, the button becomes disabled, and the label changes to reflect the fact that indexing is being stopped. 

![screen shot 2016-06-17 at 10 14 38 am](https://cloud.githubusercontent.com/assets/106961/16158933/b844a130-3474-11e6-8f6c-74a6f4b55ff5.png)

To consider:

- [ ] When *Reindex database(s)* is activated, should we warn users that reindexing might take some time, and some features of Form Runner won't be usable while reindexing happens (notably summary pages won't be usable as they won't show any information in the non-metadata columns)?
- [ ] Should summary page do something special when reindexing happens? E.g. show a message saying they are disabled? How would they detect reindexing is happening? (Is it enough to check the "static variable"? Also see 4th point below.)
- [ ] When we're stopping reindexing, instead of a button, would it better to show some text "Stopping reindexing", with a spinner next to it?
- [ ] Reindexing status is stored in a "static variable", and thus might show inconsistent information when Orbeon Forms is deployed on multiple app servers. Do we want to do something to remove this limitation? (Store information about indexing in the database? Where? In which database, if there are multiple?)

**Comment by avernet** [comment]
Left to do for reindexing UI and actions:

- [ ] Only have regular `ping-event` on the Form Runner home page in admin mode.
- [ ] On `ping-event`, we might be able to just run `recalculate`, and maybe not even that.
- [ ] Reindexing should check when the status is changed to stop, and… stop.
- [ ] Add resources for other languages.

**Comment by acspike** [comment]
Rather than making some functions unavailable during reindexing, would it be possible to generate the index into a temporary table and then drop the original and rename the temp or truncate the original table and copy the data into it after reindexing is complete? Or maybe make the index table name variable so that Orbeon can switch to a new index?

From an administrator's point of view it seems like a huge bummer to have to make a choice between a stale index and an application incapacitated for an indeterminate amount of time.

**Comment by avernet** [comment]
@acspike I agree this would be good to have, and maybe even critical for some deployments. Since we most likely won't get a chance to do something like this for 2016.2, I've created a separate issue for this: #2820. Thank you for the feedback!

**Comment by avernet** [comment]
Left for the reindexing UI:

- [ ] Change the code on the homepage to use an `xf:setvalue` instead of `calculate`, so we don't need to run a `<xf:rebuild>`, etc.
- [ ] On the homepage, don't send the ping event if we're not an admin, per the existing logic.
- [ ] Add confirmation dialog before reindexing.

**Comment by avernet** [comment]
Progress with supporting other databases, in addition to Oracle:

- [x] MySQL
    - [x] DDL
    - [x] Reindexing
    - [x] Summary page (for `a/a` and `orbeon/builder`)
    - [x] Field search
    - [x] Free text search
- [ ] SQL Server
    - [x] DDL
    - [x] Reindexing
    - [x] Summary page (for `a/a` and `orbeon/builder`)
    - [ ] Field search
    - [ ] Free text search
- [ ] DB2
    - [x] DDL
    - [ ] Reindexing
    - [ ] Summary page (for `a/a` and `orbeon/builder`)
    - [ ] Field search
    - [ ] Free text search
- [ ] PostgreSQL
    - [ ] DDL
    - [ ] Reindexing
    - [ ] Summary page (for `a/a` and `orbeon/builder`)
    - [ ] Field search
    - [ ] Free text search

**Comment by avernet** [comment]
A note about search:

- In all cases, we want it to be case insensitive.
- In all cases, I am thinking it better if we match entire words, i.e. searching "Jo" doesn't return "John". But for the free text search, on some databases, we use the rather barbaric `contains()`, which won't match entire words, so for consistency, we also don't match entire words in the field search, i.e. using `LIKE '%jo%`. This will be solved by #2830.

**Comment by avernet** [comment]
Left to do:

- [ ] Home page: when indexing is done, tell users it is so, in addition to giving the option to reindex again.
- [ ] Home page: take screenshots of the reindexing UI in different states and add the relevant explanations.
- [ ] Home page: don't send ping events if user isn't an admin, as discussed earlier.
- [ ] Code: port all code changes to the CE branch.
- [ ] Database doc: add link to new DDL, explain need to reindex after doing the changes to the tables, suggest to do it from another instance of OF deployed in parallel, and upgrading OF proper once the database has been reindexed.
- [ ] Blog: explain we added 2 index tables, how the first one makes all queries simpler and more efficient, and the second index allows producing the home page and search to be much more efficient, and easier to tune for a DBA.
- [ ] Release notes: mention change, link to blog post explaining why we did this change, and upgrade instructions.
- [ ] Customer: mention work is done, maybe after the release, is we got a chance to do all our testing.
